client/secondary_indexes.md

```diff
@@ -0,0 +1,191 @@
+
+## Overview
+
+The indexing system consists of two main components. The first is index storage, which handles storing and maintaining index information. The second is index-based document fetching, which manages retrieving documents using these indexes. Together, these components provide a robust foundation for efficient data access patterns.
+
+## Index storage
+
+### Core types
+
+The indexing system is built around several key types that define how indexes are structured and managed. At its heart is the IndexedFieldDescription, which describes a single field being indexed, including its name and whether it should be ordered in descending order. These field descriptions are combined into an IndexDescription, which provides a complete picture of an index including its name, ID, fields, and whether it enforces uniqueness.
+
+```go
+type IndexedFieldDescription struct {
+    Name string       // Field name being indexed
+    Descending bool   // Whether field is indexed in descending order
+}
+
+type IndexDescription struct {
+    Name string                      // Index name
+    ID uint32                        // Local index identifier
+    Fields []IndexedFieldDescription // Fields being indexed
+    Unique bool                      // Whether index enforces uniqueness
+}
+```
+
+The CollectionIndex interface ties everything together by defining the core operations that any index must support. This interface is implemented by different index types such as regular indexes, unique indexes, and array indexes, allowing each to provide specific behaviors while maintaining a consistent interface.
+
+```go
+type CollectionIndex interface {
+    Save(context.Context, datastore.Txn, *Document) error
+    Update(context.Context, datastore.Txn, *Document, *Document) error
+    Delete(context.Context, datastore.Txn, *Document) error
+    Name() string
+    Description() IndexDescription
+}
+```
+
+### Key structure
+
+Index keys in DefraDB follow a carefully designed format that enables efficient lookups and range scans. For regular indexes, the key format is:
+```
+<collection_id>/<index_id>(/<field_value>)+/<doc_id> -> empty value
+``` 
+Unique indexes follow a similar pattern but store the document ID as the value instead: 
+```
+<collection_id>/<index_id>(/<field_value>)+ -> <doc_id>
+```
+
+### Value encoding
+
+While DefraDB primarily uses CBOR for encoding, the indexing system employs a custom encoding/decoding solution inspired by CockroachDB. This decision was made because CBOR doesn't guarantee ordering preservation, which is crucial for index functionality. Our custom encoding ensures that numeric values maintain their natural ordering, strings are properly collated, and complex types like arrays and objects have deterministic ordering.
```

Comment:
```suggestion
DefraDB primarily uses CBOR for encoding. However, the indexing system uses a custom encoding/decoding solution, because CBOR doesn't guarantee ordering preservation, which is crucial for index functionality. 

Our custom encoding ensures that: 
- numeric values maintain their natural ordering,
- strings are properly collated, and
- complex types like arrays and objects have deterministic ordering.
```

Reply:
We might need (or want) to mention CocroachDB @jsimnz, right?